feature-BJS2-96523-test-become-a-mentor

KuaNySH88 opened the pull request.

KuaNySH88 committed:
feature-BJS2-96523-test-become-a-mentor

KuaNySH88 committed:
feature-BJS2-96523-test-become-a-mentor Fix checkstyle comments

maxmerdock reviewed (approved):
вроде всё ок

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+        long menteeId = 200L;
+        boolean validationResult;
+
+        validationResult = mentorshipValidation.canAddMentorship(mentorId,
```
Можно сразу 
`boolean validationResult = mentorshipValidation.canAddMentorship(...`

KuaNySH88 replied:
поправил

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+                (mentor, mentee) -> !Objects.equals(mentor, mentee));
+        assertTrue(validationResult);
+
+        validationResult = mentorshipValidation.canAddMentorship(mentorId,
```
Две проверки canAddMentorship надо в отдельный тест вынести. Логически они не относятся к добавлению Mentorship. В этом тесте проверяешь работу `mentorshipService.addMentorship()`, а не валидаторов как самостоятельных классов.

KuaNySH88 replied:
вынес в отдельный метод

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+                (mentor, mentee) -> !Objects.equals(mentor, mentee));
+        assertFalse(validationResult);
+
+        User mentor = new User();
```
У юзера и у других entity есть прекрасная аннотация @Builder. С помощью нее в одну строчку можно создавать объекты с сколь угодно параметрами.
`User mentor = User.builder().id(mentorId).build();`

KuaNySH88 replied:
переписал

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+        when(mentorshipRepository.getByIdOrThrow(mentorId)).thenReturn(mentor);
+        when(mentorshipRepository.getByIdOrThrow(menteeId)).thenReturn(mentee);
+
+        mentorshipService.addMentorship(mentorId, menteeId);
```
Предлагаю немного переделать метод сервиса, что бы он возвращал сохраненную сущность.
в mentorshipService надо будет лищь return добавить
`return mentorshipRepository.save(mentee);`, ну и в дто преобразовать.
Это хороший тон, когда пользователь в ответ на запрос получил сущность, которая в БД была сохранена.

Тогда и в тестах результат работы метода addMentorship можно будет проверить

KuaNySH88 replied:
добавил, еще и в deleteMentorship также добавил возвращаемое значение, правда не уверен надо ли было это делать

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+        long currentUserId = 123L;
+
+        when(userContext.getUserId()).thenReturn(currentUserId);
+        assertThrows(ForbiddenException.class,
```
Это то же отдельный кейс (отдельный тест).
Его надо вынести как в самостоятельный тест

KuaNySH88 replied:
вынес в отдельный метод

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+        mentors.add(mentor);
+        mentee.setMentors(mentors);
+
+        boolean isMentorIncluded = mentee.getMentors().contains(mentor);
```
Лучше не здесь проверку делать, а на результате, который ты на 138 строчке получил.

KuaNySH88 replied:
переделал

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+        mentor.setId(mentorId);
+        List<User> mentees = new ArrayList<>();
+        mentees.add(mentee);
+        mentor.setMentees(mentees);
```
Чуть короче можно сделать с использованием Collections.singletonList
`mentor.setMentees(Collections.singletonList(mentee);`

KuaNySH88 replied:
попробовал так сделать, но из-за того что singletonList создает immutable лист, не отработало удаление. Поэтому использовал new ArrayList<>()

KuaNySH88 committed:
feature-BJS2-96523-test-become-a-mentor
Fix PR comments
1. add return value for method addMentorship
2. add return value for method deleteMentorship
3. rewrite tests

KuaNySH88 commented:
в общем переписал полностью класс MentorshipServiceTest, изначальный вариант был ужасен

KuaNySH88 committed:
feature-BJS2-96523-test-become-a-mentor
add return value in methods in controller

SnurnicynKonstantin reviewed (approved):
А вот вейчас очень аккуратные тесты получились. Читать легко и понятно, супер!
